cloud/linode/cilium_loadbalancers.go

```diff
@@ -0,0 +1,445 @@
+		Spec: v2alpha1.CiliumLoadBalancerIPPoolSpec{
+			ServiceSelector: &slimv1.LabelSelector{
+				MatchLabels: map[string]slimv1.MatchLabelsValue{
+					"io.kubernetes.service.namespace": service.Namespace,
+					"io.kubernetes.service.name":      service.Name,
+				},
+			},
+			Blocks: []v2alpha1.CiliumLoadBalancerIPPoolIPBlock{{
+				Cidr: v2alpha1.IPv4orIPv6CIDR(fmt.Sprintf("%s/32", sharedIP)),
+			}},
+			Disabled: false,
+		},
+	}
+
+	return l.ciliumClient.CiliumLoadBalancerIPPools().Create(ctx, ciliumLBIPPool, metav1.CreateOptions{})
+}
+
+// NOTE: Cilium CRDs must be installed for this to work
+func (l *loadbalancers) deleteCiliumLBIPPool(ctx context.Context, service *v1.Service) error {
+	if err := l.retrieveCiliumClientset(); err != nil {
+		return err
+	}
+
+	return l.ciliumClient.CiliumLoadBalancerIPPools().Delete(
+		ctx,
+		fmt.Sprintf("%s-%s-pool", service.Namespace, service.Name),
+		metav1.DeleteOptions{},
+	)
+}
+
+// NOTE: Cilium CRDs must be installed for this to work
+func (l *loadbalancers) getCiliumLBIPPool(ctx context.Context, service *v1.Service) (*v2alpha1.CiliumLoadBalancerIPPool, error) {
+	if err := l.retrieveCiliumClientset(); err != nil {
+		return nil, err
+	}
+
+	return l.ciliumClient.CiliumLoadBalancerIPPools().Get(
+		ctx,
+		fmt.Sprintf("%s-%s-pool", service.Namespace, service.Name),
+		metav1.GetOptions{},
+	)
+}
+
+// NOTE: Cilium CRDs must be installed for this to work
+func (l *loadbalancers) ensureCiliumBGPPeeringPolicy(ctx context.Context) error {
+	if Options.BGPNodeSelector == "" {
+		return errNoBGPSelector
+	}
+	regionID, ok := regionIDMap[l.zone]
+	if !ok {
+		return fmt.Errorf("unsupported region for BGP: %s", l.zone)
+	}
+	if err := l.retrieveCiliumClientset(); err != nil {
+		return err
+	}
+	// check if policy already exists
+	policy, err := l.ciliumClient.CiliumBGPPeeringPolicies().Get(ctx, ciliumBGPPeeringPolicyName, metav1.GetOptions{})
+	if err != nil && !k8serrors.IsNotFound(err) {
+		klog.Infof("Failed to get CiliumBGPPeeringPolicy: %s", err.Error())
+		return err
+	}
+	// if the CiliumBGPPeeringPolicy doesn't exist, it's not nil, just empty
+	if policy != nil && policy.Name != "" {
+		return nil
+	}
+
+	// otherwise create it
+	kv := strings.Split(Options.BGPNodeSelector, "=")
+	if len(kv) != 2 {
+		return fmt.Errorf("invalid node selector %s", Options.BGPNodeSelector)
+	}
+	ciliumBGPPeeringPolicy := &v2alpha1.CiliumBGPPeeringPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ciliumBGPPeeringPolicyName,
+		},
+		Spec: v2alpha1.CiliumBGPPeeringPolicySpec{
+			NodeSelector: &slimv1.LabelSelector{MatchLabels: map[string]string{kv[0]: kv[1]}},
+			VirtualRouters: []v2alpha1.CiliumBGPVirtualRouter{{
+				LocalASN:      65001,
+				ExportPodCIDR: ptr.To(true),
+				ServiceSelector: &slimv1.LabelSelector{
+					MatchExpressions: []slimv1.LabelSelectorRequirement{{
+						Key:      "somekey",
+						Operator: slimv1.LabelSelectorOpNotIn,
+						Values:   []string{"never-used-value"},
+					}},
```

Comment:
Yes, this comes from [the cilium documentation](https://docs.cilium.io/en/stable/network/bgp-control-plane/#service-announcements) since by default, virtual routers will not announce services. 
This is a selector that matches everything since it's using "NotIn" with a dummy label.

Reply:
added